keps/sig-network/3178-iptables-cleanup/README.md

```diff
@@ -0,0 +1,687 @@
+of packet dropping in kube-proxy did it that way.)
+
+#### External Users of Kubelet's IPTables Chains
+
+Some users or third-party software may expect that certain IPTables
+chains always exist on nodes in a Kubernetes cluster.
+
+For example, the [CNI portmap plugin] provides an
+`"externalSetMarkChain"` option that is explicitly intended to be used
+with `"KUBE-MARK-MASQ"`, to make the plugin use Kubernetes's iptables
+rules instead of creating its own. ([CRI-O] also assumes that
+`KUBE-MARK-MASQ` will exist, and uses it for implementing HostPort
+mappings.) Although in most cases kube-proxy will also create
+`KUBE-MARK-MASQ`, kube-proxy may not start up until after some other
+components, and some users may be running a network plugin that has
+its own proxy implementation rather than using kube-proxy. Thus, users
+may end up in a situation where some software is trying to use a
+`KUBE-MARK-MASQ` chain that does not exist.
+
+Most of these external components have simply copied kube-proxy's use
+of `KUBE-MARK-MASQ` without understanding why it is used that way in
+kube-proxy. But because they are generally doing much less IPTables
+processing than kube-proxy does, they could fairly easily be rewritten
+to not use the packet mark, and just have slightly-redundant
+`PREROUTING` and `POSTROUTING` rules instead.
+
+[CNI portmap plugin]: https://github.com/containernetworking/plugins/tree/master/plugins/meta/portmap
+[CRI-O]: https://github.com/cri-o/cri-o
+
+#### `iptables-wrapper`
+
+Another problem is the [iptables-wrapper script] for detecting whether
+the system is using iptables-legacy or iptables-nft. Currently it
+assumes that kubelet will always have created _some_ iptables chains
+before the wrapper script runs, and so it can decide which iptables
+backend to use based on which one kubelet used. If kubelet stops
+creating chains, then iptables-wrapper may find itself in a situation
+where there are no chains in either set of tables.
+
+```
+<<[UNRESOLVED iptables-wrapper-help ]>>
+
+It might make sense to have kubelet continue to create a single
+IPTables chain that exists for no purpose other than to indicate to
+pod-based external components whether the system is using
+iptables-legacy or iptables-nft.
+
+<<[/UNRESOLVED]>>
+```
+
+[iptables-wrapper script]: https://github.com/kubernetes-sigs/iptables-wrappers
+
+## Design Details
+
+### Implementation
+
+We cannot remove `KUBE-MARK-DROP` from kubelet until we know that
+kubelet cannot possibly be running against a version of kube-proxy
+that requires it.
+
+Thus, the process will be:
+
+#### Alpha
+
+We will document, in the appropriate places, the fact that external
+components should not be assuming the existence of Kubernetes's
+IPTables chains. (This probably includes a release note and a blog
+post.)
+
+Kube-proxy will be updated to not use `KUBE-MARK-DROP`, as described
+above. (This change is unconditional; it is _not_ feature-gated.) We
```

Comment:
why no gate?  What happens if we've done something disastrously dumb - how can a cluster back out of this change?

Reply:
Since this just changes kube-proxy's internals without actually changing the visible functionality, I was considering it a cleanup/bugfix, not a feature. (So if we do something disastrously dumb, and somehow don't catch it, then affected users would have to downgrade, just like with any other major bug.)

If we feature-gate the kube-proxy change, then we'd have to push out the rest of the schedule by two releases, because you can't have the kubelet feature gate go to beta until 2 releases after kube-proxy starts not depending on `KUBE-MARK-DROP`. (So we'd have two separate feature gates, a lot like the "Move `KUBE-MARK-DROP` to Kube-Proxy Rather Than Removing It" example under "Alternatives".)